sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeUriBuilder.cs

```diff
@@ -241,6 +242,54 @@ public Uri ToUri()
             return _uri;
         }
 
+        /// <summary>
+        /// Gets the blob Uri.
+        /// </summary>
+        internal Uri ToBlobUri()
+        {
+            // AccountName cannot be null
+            // Host.Length mst be at least 5 character longth than AccountName.Length to account for ".dfs."
+            // Host must begin with "[AccountName]."
+            // The substring of Host immediately following "[AccountName]." must be "dfs."
+            if (AccountName != null
+                && Host.Length >= AccountName.Length + 5
+                && Host.Substring(0, AccountName.Length + 1).Equals(AccountName + ".", StringComparison.OrdinalIgnoreCase)
+                && Host.Substring(AccountName.Length + 1, 4).Equals(Constants.DataLake.DfsUriSuffix + ".", StringComparison.OrdinalIgnoreCase))
+            {
+                StringBuilder stringBuilder = new StringBuilder(Host);
+
+                // Replace "dfs" with "blob"
+                stringBuilder.Replace(Constants.DataLake.DfsUriSuffix, Constants.DataLake.BlobUriSuffix, AccountName.Length + 1, 3);
+                Host = stringBuilder.ToString();
+            }
+
+            return ToUri();
+        }
+
+        /// <summary>
+        /// Gets the dfs Uri.
+        /// </summary>
+        internal Uri ToDfsUri()
+        {
+            // AccontName cannot be null
+            // Host.Length must be at least 6 characters longther than AccountName.Length to account for ".blob."
+            // Host must begin with "[AccountName]."
+            // The substring of Host immediately following "[AccountName]." must be ".blob."
+            if (AccountName != null
+                && Host.Length >= AccountName.Length + 6
+                && Host.Substring(0, AccountName.Length + 1).Equals(AccountName + ".", StringComparison.OrdinalIgnoreCase)
+                && Host.Substring(AccountName.Length + 1, 5).Equals(Constants.DataLake.BlobUriSuffix + ".", StringComparison.OrdinalIgnoreCase))
+            {
+                StringBuilder stringBuilder = new StringBuilder(Host);
+
+                // Replace "blob" with "dfs"
+                stringBuilder.Replace(Constants.DataLake.BlobUriSuffix, Constants.DataLake.DfsUriSuffix, AccountName.Length + 1, 4);
+                Host = stringBuilder.ToString();
+            }
```

Comment:
Adding a separate comment on this method to highlight the whole code snippet.  We could actually use GetAccountNameFromDomain which verifies whether it's dfs or blob for you so we keep all that logic in one place:

```C#
if (!IpStyleUri)
{
    string account =  uri.GetAccountNameFromDomain(Constants.DataLake.BlobUriSuffix);
    if (account != null)
    {
        Host = Host.Replace(
            Constants.DataLake.BlobUriSuffix,
            Constants.DataLake.DfsUriSuffix,
            account.Length + 1,
            Constants.DataLake.BlobUriSuffix.Length);  
    }
}
```

It will return `null` unless it finds `"<accountname>.blob.<*>"` so it's super easy to check.

Reply:
Note that string doesn't have a Replace() method that takes 4 parameters.  I needed to use a string builder.

In addition, the internal _uri is null until ToUri() is called.  (I don't see a uri property on DataLakeUriBuilder, is this a typo?).  If I call _uri.GetAccountNameForDomain(), the result is a null pointer.

Reply:
Adding ToUri() at the beginning of this method fixes this issue.

Reply:
I think it is weird to do this as now we have it at both the top and the bottom of the same method. 
An alternative that I just talked about with Ted is pulling out the guts of GetAccountNameFromDomain into another overload like:
       
```
public static string GetAccountNameFromDomain(this Uri uri, string serviceSubDomain) =>
            GetAccountNameFromDomain(uri.Host, serviceSubDomain);

        
public static string GetAccountNameFromDomain(string host, string serviceSubDomain)
        {
            var accountEndIndex = host.IndexOf(".", StringComparison.InvariantCulture);
            if (accountEndIndex >= 0)
            {
                var serviceStartIndex = accountEndIndex + 1;
                var serviceEndIndex = host.IndexOf(".", serviceStartIndex, StringComparison.InvariantCulture);
                if (serviceEndIndex > serviceStartIndex)
                {
                    var service = host.Substring(serviceStartIndex, serviceEndIndex - serviceStartIndex);
                    if (service == serviceSubDomain)
                    {
                        return host.Substring(0, accountEndIndex);
                    }
                }
            }
            return null;
        }
```

Reply:
How does this help us not have to call ToUrl() at the top of ToDfsUrl() and ToBlobUrl()?

Reply:
We would be calling the second overload which doesn't depend on the _uri property being initialized.

Reply:
fixed.